File: Solutions/Longest-Substring-Of-All-Vowels-in-Order.py
class Solution:
    def change(self, x: str) -> int:
        if x == 'a': return 0
        elif x == 'e': return 1
        elif x == 'i': return 2
        elif x == 'o': return 3
        elif x == 'u': return 4
    def longestBeautifulSubstring(self, word: str) -> int:
        L = []
        flag = [0,0,0,0,0]
        ans = 0
        for i in word:
            if len(L) == 0 or i >= L[-1]:
                L.append(i)
                flag[self.change(i)] = 1
            else:
                if sum(flag) == 5: ans = max(ans, len(L))
                flag[0] = flag[1] = flag[2] = flag[3] = flag[4] = 0
                L.clear(), L.append(i)
                flag[self.change(i)] = 1
                
        if sum(flag) == 5: ans = max(ans, len(L))    
        return ans
    
S = Solution()
S.longestBeautifulSubstring("aeeeiiiioooauuuaeiou")
            

File: Solutions/Longest-Substring-Of-All-Vowels-in-Order.cpp
class Solution
{
  public:
    int longestBeautifulSubstring(string word)
    {
        word.push_back('a');

        int ans = 0, n = word.size(), i = 0;
        deque<pair<int, int>> stk;

        unordered_map<char, int> priority;
        for (char ch = 'a'; ch <= 'z'; ch++)
            priority[ch] = 0;
        priority['a'] = 1;
        priority['e'] = 2;
        priority['i'] = 3;
        priority['o'] = 4;
        priority['u'] = 5;

        bool flag = false;
        while (i < n)
        {
            const char &ch = word[i];
            // cout << "i: " << i << " stksize: " << stk.size() << endl;
            if (stk.empty())
            {
                if (priority[ch] <= 1)
                    stk.emplace_back(priority[ch], i);
            }
            else
            {
                if (flag)
                {
                    if (priority[ch] == 0)
                        stk.emplace_back(priority[ch], i);
                    else
                    {
                        // cout << "here" << endl;
                        ans = max(ans, stk.back().second - stk.front().second + 1);
                        flag = false;
                        stk.clear();
                        i--;
                    };
                }
                else
                {
                    if (stk.back().first == 5)
                    {
                        if (priority[ch] == 5)
                            stk.emplace_back(priority[ch], i);
                        else if (priority[ch] == 0)
                        {
                            flag = true;
                            stk.emplace_back(priority[ch], i);
                        }
                        else
                        {
                            ans = max(ans, stk.back().second - stk.front().second + 1);
                            stk.clear();
                            i--;
                        };
                    }
                    else if (stk.back().first == priority[ch] || stk.back().first + 1 == priority[ch])
                        stk.emplace_back(priority[ch], i);
                    else
                    {
                        stk.clear();
                        i--;
                    };
                };
            };
            i++;
        };
        return ans;
    }
};